### PR TITLE
Consume attestation files

### PR DESCRIPTION
**Pull request by jonasagx**
Closes: https://github.com/anchore/grype/issues/644 by consuming attestations files. Scanning attestation attached to images will be tackled later.

## output with verification
image

## output without verification
image

Please try it out:
```sh
# build local grype
go build main.go
# consume implicit attestation
./main agx-alpine-hello.att.json --key cosign.pub
# consume explicit attestation
./main att:agx-alpine-hello.att.json --key cosign.pub
```

## Design decisions
- Grype is implementing attestation verification and decoding independent of Syft, we might move them to Syft as library features used by Grype.

### CODE REVIEW

**Review by luhring — CHANGES_REQUESTED** [review]
I left some inline review comments for the code itself, but I'll also include some overall comments from looking at the feature from a user's perspective. I'll number them here so that it's a bit easier to refer to these individual comments later...

- [ ] I don't see any documentation for this new feature. We should explain how this new functionality works to users in the project's README.

- [x] It looks like piping an attestation envelope into Grype (via STDIN) doesn't work. Just like with SBOMs, users should be able to run a command like `cat ./attestation/json | grype --key ./cosign.pub` and have it behave as if the attestation path was provided as a CLI arg to Grype. We should also put this workflow under test.

- [ ] The Grype binary size increases significantly with this change — from `31 MB` to `85 MB` (almost a 3x increase). It wasn't clear to me from the description if this was known and/or expected. I know we saw a binary size increase in Syft when we introduced the attestation functionality, but I wasn't sure if the impact to Grype was known to be the exact same or if there was some mitigation opportunity available to us. (For reference, it looks like the latest Syft binary is at `67 MB`).

- [ ] I'm wondering if we considered an implementation where SBOM attestation logic lives solely within Syft and then is exported to consumers, like Grype. At first glance, this approach would seem analogous to how we handle SBOM creation and SBOM consumption between Syft and Grype, respectively. (To be clear, I don't see this as a necessity for the Grype implementation, I just want to understand the thinking.)

- [ ] Should Grype warn (or even **error out**) when the user provides a `--key` argument for an input that's **not an attestation**? I'm thinking about security expectations... Right now, if a key is provided but the input is not an attestation, Grype will continue to scan the target as usual. I think it'd be reasonable to assume that if the user has provided a `--key` argument, they are expecting a cryptographic verification to performed, and a successful verification to be required as a condition to a successful vulnerability scan. We should try to avoid scenarios where the user believes they can trust the input when really we're not sure if they can trust the input.

- [ ] When I run Grype with `att:./SOMETHING`, where the "SOMETHING" is **not an attestation**, I think the error message shown to users could be more direct about what the problem is. Right now, I see `failed to catalog: invalid payloadType  on envelope. Expected application/vnd.in-toto+json` (and actually, with an extra space after `payloadType` now that I look at the raw output again). I'm wondering if we could just say something that communicates that the input is not an attestation, even though the user told Grype to interpret the input as an attestation (using `att:`). What are your thoughts?

- [ ] When I run Grype on an attestation, but I **don't** provide the `--key` flag, I think the error message could be more direct, and could tell me that I didn't provide a public key for the verification. Right now, the error message is: `failed to catalog: failed to verify attestation signature: failed to load public key: read .: is a directory`.

**Comment by jonasagx** [comment]
Thanks for the feedback @luhring -- I am working on the actionable items. Items 3 and 4 on your list are design discussions we can have over zoom.

File: internal/config/application.go
Comment (jonasagx) [reply internal/config/application.go]: I renamed it to `AttestationKey` for now.

File: grype/pkg/syft_sbom_provider.go
Comment (jonasagx) [reply grype/pkg/syft_sbom_provider.go]: It was meant for local debugging, I just dropped it

File: grype/pkg/syft_sbom_provider.go
Comment (jonasagx) [reply grype/pkg/syft_sbom_provider.go]: yes, done

File: grype/pkg/syft_sbom_provider.go
Comment (jonasagx) [reply grype/pkg/syft_sbom_provider.go]: I thought they were closing the file, turns out the close function is not defined.
Cosign does that for some reason: https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/verify/verify.go#L125-L128

File: grype/pkg/syft_sbom_provider.go
Comment (jonasagx) [reply grype/pkg/syft_sbom_provider.go]: I refactored that function

File: grype/pkg/syft_sbom_provider.go
Comment (spiffcs) [inline grype/pkg/syft_sbom_provider.go:37]: Nice! I like that this is simpler without the extra step. A good candidate we can look at to what we might want to see eventually exported from syft since we had to write our own function here.

**Comment by jonasagx** [comment]
@luhring on binary size:
Syft's size at v0.44.0 is 70MB. 
The snapshot of Grype (Darwin/amd64) for this PR is 71MB.

File: README.md
Comment (spiffcs) [inline README.md:127]: nit:
```suggestion
Grype can scan SBOMs from attestations, as long as they are encoded as part of and [in-toto envelope](https://github.com/in-toto/attestation/blob/main/spec/README.md#envelope).
```

This should clean up the link. Let me know if you think it reads better by dropping the end statement

File: cmd/root.go
Comment (spiffcs) [inline cmd/root.go:64]: nit: instead of informing the user that the scheme is optional I would use the above help as an example and say something like `explicitly use the input as an attestation`

File: cmd/root.go
Comment (spiffcs) [inline cmd/root.go:159]: I think we want the default here to be "cosign.key" since that's the default that syft uses as well.
```suggestion
		"key", "cosign.key",
```

**Review by spiffcs — COMMENTED** [review]
Nice Job on the PR! Let's sync this afternoon about steps going forward that unblock this. I still need to read and understand the other PR comments for what else was request, but I think we're really close here!

File: grype/pkg/syft_sbom_provider.go
Comment (spiffcs) [inline grype/pkg/syft_sbom_provider.go:110]: We talked about this on the call but just wanted to capture this for others to see. I think this can return true if an sbom had some description or comment field with the mime type in it.

I think we want to make sure we account for cases where `isPossibleAttestation` and `isPossibleSBOM` both return true. 

If both return true then only one of the resulting workflow will not return an error so we can try both and decide which one to return to the user. If both return an error saying we tried both

File: test/cli/utils_test.go
Comment (spiffcs) [inline test/cli/utils_test.go:125]: Was this erroring before? What was the reason for the change in this file?

**Comment by spiffcs** [comment]
Small comments added and I think this needs a rebase or merge into master

File: test/cli/utils_test.go
Comment (jonasagx) [reply test/cli/utils_test.go:125]: this approach errors when the input file is > 30k, I suspect the stdin buffer got full making the test hang for ever.

**Comment by spiffcs** [comment]
@jonasagx this looks really good! Nice job on the PR and thanks for responding to the feedback. 

All my changes have been addressed. Waiting for @luhrings feedback. Also maybe we can all sync on the concerns about the size?

**Comment by spiffcs** [comment]
Per meeting comments - We need to update the error for when an attestation prefix is passed on a file that is a regular `JSON` sbom. The test should also be updated to check the error text.

File: grype/pkg/syft_sbom_provider.go
Comment (spiffcs) [inline grype/pkg/syft_sbom_provider.go:61]: I think we don't want to change the signature here. `getSBOMReader` should not need to return the extra info here . `getSBOMReader` can be given to the agency to determine which error it returns based on the input discovered.

File: grype/pkg/syft_sbom_provider_test.go
Comment (spiffcs) [inline grype/pkg/syft_sbom_provider_test.go:35]: Per meeting comments `fixture` => `input`

File: grype/pkg/syft_sbom_provider_test.go
Comment (spiffcs) [inline grype/pkg/syft_sbom_provider_test.go:106]: replace all instances of mistake `schema` => `scheme`

File: README.md
Comment (spiffcs) [inline README.md:127]: nit: I think we remove comma here

Comment (spiffcs) [reply README.md:127]: Apologies if this came from my earlier suggestion

**Comment by spiffcs** [comment]
Fast follow after this PR on validating predicate type in case of modification between attestation generation and attestation consumption

File: README.md
Comment (spiffcs) [reply README.md:127]: Comment from meeting - Add information to feature session for this PR

Comment (spiffcs) [reply README.md:127]: Comment from meeting - update scheme section to talk about new source and how it is a more secure version of the grype workflow

Comment (spiffcs) [reply README.md:127]: Comment from meeting - update Scan attestations section Examples with full connective story of commands

**Comment by spiffcs** [comment]
From Meeting: Update TUI message to show verification rather than bury it in the info section

File: internal/config/application.go
Comment (luhring) [reply internal/config/application.go]: Thought: should we update to `AttestationVerificationKey` or something like that? We're considering adding attestation creation down the road, where we'd need a signing key, too. At that point, I could see `AttestationKey` being ambiguous as to its intended use... WDYT?

File: grype/pkg/syft_sbom_provider.go
Comment (luhring) [reply grype/pkg/syft_sbom_provider.go:101]: @jonasagx Ping on this one as still needing an answer

File: internal/config/application.go
Comment (luhring) [reply internal/config/application.go]: A related question: Should `Attestation` be its own config section?

File: grype/pkg/syft_sbom_provider.go
Comment (jonasagx) [reply grype/pkg/syft_sbom_provider.go:101]: With attestations two condition were added, and the conditions are all mutually exclusive so it seems easier to read them as a switch/case block.

File: internal/config/application.go
Comment (jonasagx) [reply internal/config/application.go]: 1. The field name currently is `AttestationKey`, which doesn't differentiate whether it is for public our private key, so we can use the same field for verification _and_ creation of an attestation. This is similar to how [cosign handles](https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/attest.go#L67) the `--key` parameter.
2. Adding an attestation section sounds like a good idea, but I for the use cases of this PR the only field needed is a key. We can include it now and expand later as well. Keyless workflows might be the biggest contributor here -- cc @spiffcs

Comment (luhring) [reply internal/config/application.go]: When it comes to code, I think we should be careful not to use the same data field to store something that's not sensitive (like a public key) as something that is sensitive (like a private key). The consequence of getting these two keys mixed up could be very, very bad.

>so we can use the same field for verification and creation of an attestation

I don't think I follow... How would this work? If Grype was consuming an SBOM attestation and producing a vulnerability scan attestation at the same time, we'd use the same key parameter for both the public verification key and the private signing key?

File: ui/event_handlers.go
Comment (jonasagx) [inline ui/event_handlers.go:175]: @luhring WDYT of this message?

File: ui/event_handlers.go
Comment (jonasagx) [inline ui/event_handlers.go:28]: @wagoodman is this ok? Background: increased it so new (longer) messages get aligned.

File: internal/config/application.go
Comment (jonasagx) [reply internal/config/application.go]: I see, you're right. I will rename the config field to `Attestation.VerificationKey`, since it has its own section now.

File: ui/event_handlers.go
Comment (luhring) [reply ui/event_handlers.go:175]: @jonasagx I'd say "verification" instead of "proof"

File: ui/event_handlers.go
Comment (luhring) [inline ui/event_handlers.go:157]: I don't think we need to output the signature itself to the TUI

File: ui/event_handlers.go
Comment (luhring) [inline ui/event_handlers.go:176]: I don't think we need this secondary message

File: ui/event_handlers.go
Comment (jonasagx) [reply ui/event_handlers.go]: please ignore

**Review by spiffcs — APPROVED** [review]
Approved with new changes thanks @jonasagx - Looking for @luhring review and to remove his request for changes here

**Review by luhring — APPROVED** [review]

**Pull request merged**
